lib/astc_encode.cpp

```diff
@@ -609,7 +610,7 @@ ktxTexture2_CompressAstcEx(ktxTexture2* This, ktxAstcParams* params) {
                         block_size_x, block_size_y, block_size_z);
     quality = astcQuality(params->qualityLevel);
     profile = astcEncoderAction(*params, BDB);
-    swizzle = astcSwizzle(*params);
+    swizzle = params->normalMap ? normal_swizzle : astcSwizzle(*params);
```

Comment:
The Basis encoder uses `inputSwizzle` from params if set, otherwise it uses the default normal_map swizzle. I think this encoder should do the same.

Reply:
Right, this is why I had this in `toktx.cc`. We had this conversation before that I can't find in the resolved conversation now. The problem was in the enocder I can't tell if its "inputSwizzle" or "defaultSwizzle".  Because if inputSwizzle wasn't defined, it was set to default swizzle unconditionally.

Reply:
`defaultSwizzle` in `toktx` is only used when we need to convert a 1 or 2 component image to luminance. (This happens when the input file spec defines such 1 or 2 component files as luminance which PNG does.) If `options.inputSwizzle` is set that is copied to the params struct instead of `defaultSwizzle`. If neither is set, nothing is passed. At least that is what should happen.

Within `basis_encode.cpp` I use `params->inputSwizzle`, if set, otherwise various preset swizzles depending on the number of channels in the input and whether `normal_map` is set. In this case I need to determine if we are making a luminance texture because the DFD channel id(s) are set differently for luminance vs. actual splitting of r & g. I'm thinking of adding a new field to `ktxBasisParams` for the caller to tell me it is a luminance texture rather than trying to guess as I am now. The channel id(s) are used to guide the choice of transcode target. The ASTC encoder does not have to worry about this as there is only a single channel id.

I don't think you need to tell if what you are receiving was `toktx's` `inputSwizzle` or its `defaultSwizzle`. They're both just the `inputSwizzle` you are receiving in `params`. If the caller sets a stupid swizzle for the `normal_map` case that is their problem. I want to provide the flexibility for people to set their own.

I think the ASTC encoder should always use `inputSwizzle`, if set and have similar code to that in lines 529 to 540 of basis_encode.cpp for setting a swizzle when it is not.

Reply:
The issue I see is that `toktx` is passing `defaultSwizzle` as the `inputSwizzle` to the encoders, if it is set. It should do that only if `--normal_mode` is not specified. If `--normal_mode` is specified and the input has 2 components then clearly the user is not making a luminance texture.

Reply:
what do you think about https://github.com/KhronosGroup/KTX-Software/pull/493/commits/d10c7ce79b11b548b9cdc36bfd9876e7a1595724?

Reply:
Looks good to me. Now we need to figure out why that one test is failing on Linux and Windows but not on macOS.